internal/kademlia/node.go

```diff
@@ -140,31 +136,41 @@ func (kademlia *Node) FindValue(hash string) string {
 		// a list of nodes to know which nodes has been probed already
 		probedNodes := ContactCandidates{}
 
-		for {
-			updateClosest := false
-			numProbed := 0
+		return kademlia.findValueAmongContacts(hash, alpha, shortList, currentClosest, currentClosest.distance, probedNodes)
+	}
+}
 
-			for i := 0; i < shortList.Len() && numProbed < alpha; i++ {
+func (kademlia *Node) findValueAmongContacts(hash string, alpha int, shortList ContactCandidates, currentClosest Contact, distance *NodeID, probedNodes ContactCandidates) string {
+	for {
+		updateClosest := false
+		numProbed := 0
+
+		for i := 0; i < shortList.Len() && numProbed < alpha; i++ {
+
+			if probedNodes.Contains(shortList.contacts[i]) {
+				continue
+			} else {
+				rpc, err := kademlia.network.SendFindDataMessage(&shortList.contacts[i], &kademlia.RT.me, hash)
 
-				if probedNodes.Contains(shortList.contacts[i]) {
+				if *rpc.Payload.Value != "" {
+					return *rpc.Payload.Value
+				}
+
+				// if a node responds with an error remove that node
+				// from the shortlist and from the bucket
+				if err != nil {
+					log.Warn(err)
+					kademlia.RT.RemoveContact(shortList.contacts[i])
+					shortList.contacts = append(shortList.contacts[:i], shortList.contacts[i+1:]...)
```

Comment:
after the check on rpc.Payload.Value the method does the same thing as in findNodeAmongContacts, can this be refactored into a different function as well to remove the duplicate code?